### DISCUSSION TITLE
Support for inline code syntax highlighting

### DISCUSSION THREAD

**Discussion by jktrn** [expressive-code/expressive-code#395] [Feature Requests]
Hello!

I was wondering if it was possible for syntax highlighting of inline code blocks, e.g. some form of syntax along the lines of `const foo = bar{:js}` which would highlight the block in the user's selected theme. As of now, this is supported in:
- [`@shikijs/rehype`](https://shiki.matsu.io/packages/rehype#inline-code), but attempting to import this overrides expressive-code's `<pre>`-based code blocks
- [`rehype-pretty-code`](https://rehype-pretty.pages.dev/) which ships it out-of-the-box

Could this be possible within expressive-code with some form of custom Shiki transformer or any other means? Thank you for your time.

**Comment by hippotastic** [comment]
Hello and welcome! :) Thank you for this feature request!

This is currently not possible in Expressive Code, I'd need to add support in the plugin API and all built-in plugins for it so that they can distinguish between block-level code processing and inline code processing. In inline code, most rendering features like frames, line numbers, collapsible blocks etc. would not be possible because no block-level tags are allowed.

I'm honestly not very happy with the `{:js}` syntax, but if Shiki establishes this as a standard, I'd probably use the same syntax as well to avoid fragmentation in the syntax highlighting landscape.

That being said, I'm not sure yet how useful it actually is to highlight inline code. Do you have any real-world examples I could look at where inline highlighting is being used and looks helpful?

**Comment by jktrn** [comment]
> In inline code, most rendering features like frames, line numbers, collapsible blocks etc. would not be possible because no block-level tags are allowed.

I think that if you're at the point where you would need to add collapsible blocks, frames, or line numbers to an inline block it might as well be converted into a `<pre>`, so this should not be a problem!

> I'm honestly not very happy with the `{:js}` syntax, but if Shiki establishes this as a standard, I'd probably use the same syntax as well to avoid fragmentation in the syntax highlighting landscape.

I unfortunately don't like it that much either, but it seems like this syntax (or variations of it) are an established thing already in a couple of examples I found:
- [kramdown](https://kramdown.gettalong.org/syntax.html#span-ials)
- [pandoc](https://pandoc.org/MANUAL.html#extension-inline_code_attributes)
- [mayurankv/Obsidian-Code-Styler](https://github.com/mayurankv/Obsidian-Code-Styler#inline-code) (which does its own thing with curly brackets, but also supports pandoc notation)

The main maintainer for Shiki had this discussion [here](https://github.com/shikijs/shiki/issues/660#issuecomment-2069110522) and a couple of good points were made, including that nesting the curly brackets within the backticks themselves (`` `code{:js}` ``) is better than putting it outside `` `code`{:js} `` since the latter creates ambiguity within MDX files, and I believe the colon is there to help the parser discern between code snippets involving curly brackets. Some inline code parsers support VSCode token syntax like `{:.entity.name.function}` for arbitrary coloring based on a user-supplied VSCode theme's JSON.

> That being said, I'm not sure yet how useful it actually is to highlight inline code. Do you have any real-world examples I could look at where inline highlighting is being used and looks helpful?

This feature itself is very, very uncommon--hence why supporting it would make documentation significantly more polished (especially in downstreams e.g. Starlight). These aren't considered "real-life examples" since they are documentation for already existing implementations of this feature (and or my own creations), but I hope you can see the appeal:

- From the [rehype-pretty-code](https://rehype-pretty.pages.dev/) docs: ![image](https://github.com/user-attachments/assets/09b72fae-afa0-4ed5-95d1-ba184e06b21c)
- From the [mayurankv/Obsidian-Code-Styler](https://github.com/mayurankv/Obsidian-Code-Styler#inline-code) docs: ![image](https://github.com/user-attachments/assets/bd75a5dc-b975-48b7-a326-4eb02ed63d10)
- From my [personal blogging template](https://astro-erudite.vercel.app/blog/the-state-of-static-blogs): 
![image](https://github.com/user-attachments/assets/df387b5d-ae8e-40a8-bb4e-8f3f3af1c0e9)
![image](https://github.com/user-attachments/assets/83c92508-163f-49a6-8811-0cf03e14d438)

Inline syntax highlighting would be useful in these particular use cases, taken from the expressive-code docs itself:
- ![image](https://github.com/user-attachments/assets/e667e618-e6a5-4d77-905a-534bcbcac181)
- ![image](https://github.com/user-attachments/assets/33612a83-23e7-470e-b97e-53b939b50c0c)
- ![image](https://github.com/user-attachments/assets/bc2467c0-5b0b-4bdf-86e5-01e163c0e86a)

The point of syntax highlighting itself is to help the human eye break up tokens into recognizable segments, which is especially true for longer segments e.g. the examples from the docs above. Hopefully this clears things up!

**Comment by lloydjatkinson** [comment]
I found an example of this here https://rehype-pretty.pages.dev/.

![image](https://github.com/user-attachments/assets/86741bd7-8bea-4dbe-93e4-98c54ea05b8e)

I think it would probably be something people would want to be supported but default off? Not sure.

Also, I think personally that grey background only works when you *don't* have colour as that contrast doesn't look strong enough. So, there's things to think about I suppose.

**Comment by jktrn** [comment]
Hi @hippotastic, is this in your pipeline/planned at all? I can attempt to take a look and send a PR if desired.

**Comment by hippotastic** [comment]
> Hi [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/6137925?u=b3e6ae058335207e17e3309c57805b98a4170fd2&amp;v=4&amp;size=80">@hippotastic](https://github.com/hippotastic?rgh-link-date=2025-01-25T01%3A38%3A26.000Z), is this in your pipeline/planned at all? I can attempt to take a look and send a PR if desired.

Hi @jktrn, yes, basically every feature request I leave open is something I can imagine having in Expressive Code. I liked your idea and also gave your previous post outlining the idea in more detail a thumbs up. I just didn't get to it yet. :)

Thank you very much for offering to implement this! Before you attempt to do it, I'd like to come back to this point I mentioned in my first response:

> I'd need to add support in the plugin API and all built-in plugins for it so that they can distinguish between block-level code processing and inline code processing. In inline code, most rendering features like frames, line numbers, collapsible blocks etc. would not be possible because no block-level tags are allowed.

In my opinion, getting this right in terms of the plugin API is the most crucial part of this feature request. We need to find a way to extend the existing plugin API so that plugins can somehow perform work on inline code as well, while making sure that all existing plugins (which are not aware of this new feature) don't get called on inline code while continuing to work as normal on regular code blocks.

You can take my existing plugins like frames, collapsible sections and line numbers as examples for your testing. Imagine that these were third-party code you cannot modify, and try to come up with an extension to the existing plugin API that adds inline code support without breaking these existing plugins. Without having to do any modifications to them, they should not get called on inline code, while your new version of `@expressive-code/plugin-shiki` should be able to declare somehow that it supports inline code. In its hooks, it also needs to have a way of recognizing if the code it's currently processing is inline code or a regular code block, in order to only process the special curly brace language suffix in inline code.

I'd be happy to discuss the API idea here first, but if you want to try something out first (or even already have an idea), feel free to risk it and and show a possible implementation in a PR right away. Either way, thanks again for your offer!

**Comment by techfg** [comment]
I've been looking to finding a solution for this as I agree it's a highly valuable feature.

Upon doing so, some initial testing shows that you can utilize `expressive-code` (for code blocks) alongside `rehype-shiki` or `rehype-pretty-code` (for inline code).  The key is to ensure that `expressive-code` comes before the others in the list of rehype plugins.  

The reason this appears to work is as follows:

1. EC will eliminate the `language-*` from the code block class names during its processing
2. EC will ignore any inline code blocks
3. Both [rehype-shiki](https://github.com/shikijs/shiki/blob/ac6bea51b807a82e0a5c9a83796c3abdda0fc293/packages/rehype/src/core.ts#L129) and [rehype-pretty-code](https://github.com/rehype-pretty/rehype-pretty-code/blob/eba61cbe7cb22a230702ff1e20b4483b917ae220/packages/core/src/index.ts#L357) will ignore any code blocks where it cannot find a `language-*` class name

I have created a [demo](https://stackblitz.com/edit/sb1-95yslf6f?file=js%2Fmain.js) of using side-by-side, see example screenshots from the demo below.

There are a few drawbacks to this approach:
1. Depending on your EC theme, the theme used for inline blocks will be different
2. Even if your EC theme is a Shiki theme, you'll end up with multiple shiki highlighters.  Both `rehype-shiki` and `rehype-pretty-code` support an option for providing a highlighter directly so if/when #335 is merged, you would be able to utilize the same highlighter across both plugins you are using

Of course, having two code block processors isn't ideal and it would be preferrable to have one that handles both block code and inline code.

Along that line, I have submitted #336, would welcome any feedback/thoughts!

## Demo Screenshots of using multiple plugins for inline code highlighting 
[Demo](https://stackblitz.com/edit/sb1-95yslf6f?file=js%2Fmain.js) - Please be patient upon initial startup, may take a minute even after initial webpage displays.

### Expressive Code Only
![Image](https://github.com/user-attachments/assets/92de631c-2224-48a7-9edf-75bbbd00b514)

### Expressive Code w/ Rehype-Shiki
![Image](https://github.com/user-attachments/assets/04a65f5c-28cd-4965-9187-e4c20c5f2ec4)

### Expressive Code w/ Rehype-Pretty-Code
![Image](https://github.com/user-attachments/assets/0a2e0266-afbd-402d-b148-8dfb8bd6bf15)

**Comment by jktrn** [comment]
@hippotastic Hi, there's a PR for this now! Any chance you can take a look?

**Comment by delucis** [comment]
Hello! We are moving feature requests and support questions to discussions and using issues only for bugs to help better track those going forward. See you in the [Discussions](https://github.com/expressive-code/expressive-code/discussions) tab!

**Comment by babakfp** [comment]
Any news?

**Reply by babakfp** [reply]
shiki and rehype-pretty-code supports this. things just need to get updated I guess. i would expect astro itself to support this but I don't know what is going on their side to not support it yet!

**Reply by jktrn** [reply]
I've ended up just writing a plugin for this because they are taking a while

See https://github.com/jktrn/astro-erudite/blob/main/src%2Flib%2Fexpressive-code%2Finline.ts

**Reply by delucis** [reply]
Nice! If you’d like to publish this to npm you could add it to the [Community Plugins](https://expressive-code.com/plugins/community-plugins/) page @jktrn.
